stdlib/public/Concurrency/TaskPrivate.h

```diff
@@ -787,45 +947,110 @@ inline void AsyncTask::flagAsAndEnqueueOnExecutor(ExecutorRef newExecutor) {
     // We can only be suspended if we were previously running. See state
     // transitions listed out in Task.h
     assert(oldStatus.isRunning() && !oldStatus.isEnqueued());
+    taskWasEscalated = oldStatus.isStoredPriorityEscalated();
+    taskStoredPriority = oldStatus.getStoredPriority();
 
-    newStatus = oldStatus.withRunning(false);
+    newStatus = newStatus.withRunning(false);
     newStatus = newStatus.withoutStoredPriorityEscalation();
+    newStatus = newStatus.withTaskDependency();
+
+    // Escalate the thing we are dependent on to have the max priority that we
+    // see at the time of publishing the record. If there is another concurrent
+    // escalator of the task, we will still converge since escalation is a
+    // maxing function and the other escalator will end up escalating the
+    // dependency record we publish here.
+    //
+    // Note that we have to do this escalation while adding the status record
+    // and not after - we are not guaranteed to be able to have a valid
+    // reference to the dependencyStatusRecord or its contents, once we have
+    // published it in the ActiveTaskStatus.
+    SWIFT_TASK_DEBUG_LOG("[Dependency] Escalate the dependency %p of task %p",
+                  dependencyStatusRecord, this);
+    dependencyStatusRecord->performEscalationAction(taskStoredPriority);
+
+    // Always add the dependency status record
+    return true;
+  });
 
-    if (_private()._status().compare_exchange_weak(oldStatus, newStatus,
-            /* success */std::memory_order_relaxed,
-            /* failure */std::memory_order_relaxed)) {
-      break;
-    }
-  }
 #if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
   // Successfully dropped task drain lock, make sure to remove override on
   // thread due to task
-  if (oldStatus.isStoredPriorityEscalated()) {
+  if (taskWasEscalated) {
      SWIFT_TASK_DEBUG_LOG("[Override] Reset override %#x on thread from task %p",
-       oldStatus.getStoredPriority(), this);
-     swift_dispatch_lock_override_end((qos_class_t) oldStatus.getStoredPriority());
+       taskStoredPriority, this);
+     swift_dispatch_lock_override_end((qos_class_t) taskStoredPriority);
   }
 #endif
 
-  newStatus.traceStatusChanged(this);
   swift_task_exitThreadLocalContext((char *)&_private().ExclusivityAccessSet[0]);
   restoreTaskVoucher(this);
   return;
 }
 
+inline void AsyncTask::destroyTaskDependency(TaskDependencyStatusRecord *dependencyRecord) {
+  assert(_private().dependencyRecord == dependencyRecord);
+
+  dependencyRecord->destroy();
+  _swift_task_dealloc_specific(this, dependencyRecord);
+
+  _private().dependencyRecord = nullptr;
+}
+
+// this -> task which is suspending
+// Input task -> task we are waiting on
+inline void AsyncTask::flagAsSuspendedOnTask(AsyncTask *task) {
+  assert(_private().dependencyRecord == nullptr);
+
+  void *allocation = _swift_task_alloc_specific(this, sizeof(class TaskDependencyStatusRecord));
```

Comment:
It would be great if this was just part of the task structure itself and not something we allocated on the fly. 

cc: @rjmccall

Reply:
We could definitely just have stable space for this in the task header, as long as we don't touch it as soon as we remove it.